### PR TITLE
testing: add monkey test

### PR DESCRIPTION
**Pull request by FASTSHIFT**
## Summary
A UI stress test program based on uinput.
The currently implemented functions are:
1. Support random click and swipe
2. Support recording user operations and playback
3. Support multiple instances, you can start multiple monkey processes to run independently at the same time
4. Support simultaneous monitoring and operation of multiple devices

## Impact
Add new features.

## Testing
daily test.

### CODE REVIEW

**Comment by xiaoxiang781216** [comment]
@FASTSHIFT please fix the warning report here: 
https://github.com/apache/incubator-nuttx-apps/runs/7842281876?check_suite_focus=true

**Comment by FASTSHIFT** [comment]
> @FASTSHIFT please fix the warning report here: https://github.com/apache/incubator-nuttx-apps/runs/7842281876?check_suite_focus=true

done

**Review by xiaoxiang781216 — APPROVED** [review]

File: testing/monkey/monkey.c [hunk 1]
```diff
@@ -0,0 +1,196 @@
...
+    {                                                                     \
+      FAR struct monkey_dev_s *dev;                                       \
+      dev = monkey_dev_create(CONFIG_TESTING_MONKEY_DEV_PATH_##type,      \
+                              MONKEY_DEV_TYPE_##type);                    \
+      if (!dev)                                                           \
+        {                                                                 \
+          goto failed;                                                    \
+        }                                                                 \
+      monkey->devs[monkey->dev_num] = dev;                                \
+      monkey->dev_num++;                                                  \
```
Comment (pkarashchenko) [inline testing/monkey/monkey.c:52]:
```suggestion
      (monkey)->devs[(monkey)->dev_num] = dev;                            \
      (monkey)->dev_num++;                                                \
```

File: testing/monkey/monkey.c [hunk 2]
```diff
@@ -0,0 +1,196 @@
...
+#include "monkey_dev.h"
+#include "monkey_utils.h"
+
+/****************************************************************************
+ * Pre-processor Definitions
+ ****************************************************************************/
+
+#define MONKEY_DEV_CREATE_MATCH(monkey, type_mask, type)                  \
+do {                                                                      \
+  if ((type_mask & MONKEY_DEV_TYPE_##type) == MONKEY_DEV_TYPE_##type)     \
```
Comment (pkarashchenko) [inline testing/monkey/monkey.c:42]:
```suggestion
  if (((type_mask) & MONKEY_DEV_TYPE_##type) == MONKEY_DEV_TYPE_##type)   \
```

File: testing/monkey/monkey.c [hunk 3]
```diff
@@ -0,0 +1,196 @@
...
+
+/****************************************************************************
+ * Name: monkey_create
+ ****************************************************************************/
+
+FAR struct monkey_s *monkey_create(int dev_type_mask)
+{
+  FAR struct monkey_s *monkey = malloc(sizeof(struct monkey_s));
+  MONKEY_ASSERT_NULL(monkey);
+  memset(monkey, 0, sizeof(struct monkey_s));
```
Comment (pkarashchenko) [inline testing/monkey/monkey.c:68]: Optional: `zalloc` or `calloc` can used to remove `memset`

File: testing/monkey/monkey.c [hunk 4]
```diff
@@ -0,0 +1,196 @@
...
+ ****************************************************************************/
+
+void monkey_delete(FAR struct monkey_s *monkey)
+{
+  int i;
+  MONKEY_ASSERT_NULL(monkey);
+
+  for (i = 0; i < monkey->dev_num; i++)
+    {
+        monkey_dev_delete(monkey->devs[i]);
```
Comment (pkarashchenko) [inline testing/monkey/monkey.c:109]:
```suggestion
      monkey_dev_delete(monkey->devs[i]);
```

File: testing/monkey/monkey.h [hunk 5]
```diff
@@ -0,0 +1,98 @@
...
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
+ * WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.  See the
+ * License for the specific language governing permissions and limitations
+ * under the License.
+ *
+ ****************************************************************************/
+
+#ifndef __MONKEY_H__
+#define __MONKEY_H__
```
Comment (pkarashchenko) [inline testing/monkey/monkey.h:22]: please include path part

File: testing/monkey/monkey_assert.h [hunk 6]
```diff
@@ -0,0 +1,37 @@
...
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
+ * WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.  See the
+ * License for the specific language governing permissions and limitations
+ * under the License.
+ *
+ ****************************************************************************/
+
+#ifndef __MONKEY_ASSERT_H__
+#define __MONKEY_ASSERT_H__
```
Comment (pkarashchenko) [inline testing/monkey/monkey_assert.h:22]: Please include path part

File: testing/monkey/monkey_dev.c [hunk 7]
```diff
@@ -0,0 +1,327 @@
...
+{
+  FAR struct monkey_dev_s *dev;
+  int oflag;
+  int fd;
+
+  MONKEY_ASSERT_NULL(dev_path);
+
+  dev = malloc(sizeof(struct monkey_dev_s));
+  MONKEY_ASSERT_NULL(dev);
+  memset(dev, 0, sizeof(struct monkey_dev_s));
```
Comment (pkarashchenko) [inline testing/monkey/monkey_dev.c:152]: Optional: `zalloc` or `calloc`

File: testing/monkey/monkey_dev.c [hunk 8]
```diff
@@ -0,0 +1,327 @@
...
+    {
+      free(dev);
+    }
+
+  return NULL;
+}
+
+/****************************************************************************
+ * Public Functions
+ ****************************************************************************/
```
Comment (pkarashchenko) [inline testing/monkey/monkey_dev.c:194]: `monkey_dev_create` also seems to be `Public`

File: testing/monkey/monkey_dev.c [hunk 9]
```diff
@@ -0,0 +1,327 @@
...
+      /* Reset input state */
+
+      struct monkey_dev_state_s state;
+      memset(&state, 0, sizeof(state));
+      state.type = dev->type;
+      monkey_dev_set_state(dev, &state);
+
+      MONKEY_LOG_NOTICE("close fd: %d", dev->fd);
+      close(dev->fd);
+      dev->fd = -1;
```
Comment (pkarashchenko) [inline testing/monkey/monkey_dev.c:215]: not needed since is followed by `free` unconditionally

File: testing/monkey/monkey_dev.c [hunk 10]
```diff
@@ -0,0 +1,327 @@
...
+
+  free(dev);
+}
+
+/****************************************************************************
+ * Name: monkey_dev_set_state
+ ****************************************************************************/
+
+void monkey_dev_set_state(FAR struct monkey_dev_s *dev,
+                           FAR const struct monkey_dev_state_s *state)
```
Comment (pkarashchenko) [inline testing/monkey/monkey_dev.c:226]:
```suggestion
                          FAR const struct monkey_dev_state_s *state)
```

File: testing/monkey/monkey_dev.c [hunk 11]
```diff
@@ -0,0 +1,327 @@
...
+                   state->data.touch.y,
+                   state->data.touch.is_pressed);
+      break;
+
+    case MONKEY_DEV_TYPE_BUTTON:
+      ubutton_write(dev->fd, state->data.button.value);
+      break;
+
+    default:
+      break;
```
Comment (pkarashchenko) [inline testing/monkey/monkey_dev.c:244]: I think 2 more spaces are needed for case block

File: testing/monkey/monkey_dev.c [hunk 12]
```diff
@@ -0,0 +1,327 @@
...
+                          &state->data.touch.is_pressed);
+      break;
+
+    case MONKEY_DEV_TYPE_BUTTON:
+      retval = button_read(dev->fd, &state->data.button.value);
+      break;
+
+    default:
+      break;
+    }
```
Comment (pkarashchenko) [inline testing/monkey/monkey_dev.c:277]: ditto

File: testing/monkey/monkey_dev.h [hunk 13]
```diff
@@ -0,0 +1,97 @@
...
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
+ * WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.  See the
+ * License for the specific language governing permissions and limitations
+ * under the License.
+ *
+ ****************************************************************************/
+
+#ifndef __MONKEY_DEV_H__
+#define __MONKEY_DEV_H__
```
Comment (pkarashchenko) [inline testing/monkey/monkey_dev.h:22]: ditto

File: testing/monkey/monkey_log.c [hunk 14]
```diff
@@ -0,0 +1,99 @@
...
+                       FAR const char *func,
+                       FAR const char *fmt,
+                       ...)
+{
+  struct va_format vaf;
+  va_list ap;
+
+  static const int priority[_MONKEY_LOG_LEVEL_LAST] = {
+    LOG_INFO, LOG_NOTICE, LOG_WARNING, LOG_ERR
+  };
```
Comment (pkarashchenko) [inline testing/monkey/monkey_log.c:61]:
```suggestion
  static const int priority[_MONKEY_LOG_LEVEL_LAST] =
    {
      LOG_INFO, LOG_NOTICE, LOG_WARNING, LOG_ERR
    };
```

File: testing/monkey/monkey_log.h [hunk 15]
```diff
@@ -0,0 +1,117 @@
...
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
+ * WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.  See the
+ * License for the specific language governing permissions and limitations
+ * under the License.
+ *
+ ****************************************************************************/
+
+#ifndef __MONKEY_LOG_H__
+#define __MONKEY_LOG_H__
```
Comment (pkarashchenko) [inline testing/monkey/monkey_log.h:22]: ditto

File: testing/monkey/monkey_main.c [hunk 16]
```diff
@@ -0,0 +1,381 @@
...
+ * Pre-processor Definitions
+ ****************************************************************************/
+
+#define MONKEY_PREFIX "monkey"
+
+#define OPTARG_TO_VALUE(value, type, base)                             \
+  do                                                                   \
+  {                                                                    \
+    FAR char *ptr;                                                     \
+    value = (type)strtoul(optarg, &ptr, base);                         \
```
Comment (pkarashchenko) [inline testing/monkey/monkey_main.c:46]:
```suggestion
    (value) = (type)strtoul(optarg, &ptr, (base));                     \
```

File: testing/monkey/monkey_recorder.c [hunk 17]
```diff
@@ -0,0 +1,344 @@
...
+ ****************************************************************************/
+
+void monkey_recorder_delete(FAR struct monkey_recorder_s *recorder)
+{
+  MONKEY_ASSERT_NULL(recorder);
+  if (recorder->fd > 0)
+    {
+      MONKEY_LOG_NOTICE("close fd: %d", recorder->fd);
+      close(recorder->fd);
+      recorder->fd = -1;
```
Comment (pkarashchenko) [inline testing/monkey/monkey_recorder.c:201]:
```suggestion
```
ditto

File: testing/monkey/monkey_recorder.c [hunk 18]
```diff
@@ -0,0 +1,344 @@
...
+                         &state->data.touch.y,
+                         &state->data.touch.is_pressed);
+      if (converted != 5)
+        {
+          return MONKEY_RECORDER_RES_PARSER_ERROR;
+        }
+      break;
+    case MONKEY_DEV_TYPE_BUTTON:
+      converted = sscanf(buffer,
+                        MONKEY_REC_BTN_FMT,
```
Comment (pkarashchenko) [inline testing/monkey/monkey_recorder.c:318]: 2 more spaces for case block

File: testing/monkey/monkey_recorder.h [hunk 19]
```diff
@@ -0,0 +1,110 @@
...
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
+ * WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.  See the
+ * License for the specific language governing permissions and limitations
+ * under the License.
+ *
+ ****************************************************************************/
+
+#ifndef __MONKEY_RECORDER_H__
+#define __MONKEY_RECORDER_H__
```
Comment (pkarashchenko) [inline testing/monkey/monkey_recorder.h:22]: ditto

File: testing/monkey/monkey_type.h [hunk 20]
```diff
@@ -0,0 +1,118 @@
...
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
+ * WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.  See the
+ * License for the specific language governing permissions and limitations
+ * under the License.
+ *
+ ****************************************************************************/
+
+#ifndef __MONKEY_TYPE_H__
+#define __MONKEY_TYPE_H__
```
Comment (pkarashchenko) [inline testing/monkey/monkey_type.h:22]: ditto

File: testing/monkey/monkey_type.h [hunk 21]
```diff
@@ -0,0 +1,118 @@
...
+#include <stdint.h>
+#include <stdbool.h>
+
+/****************************************************************************
+ * Pre-processor Definitions
+ ****************************************************************************/
+
+#define MONKEY_UINPUT_TYPE_MASK     (0x10)
+#define MONKEY_IS_UINPUT_TYPE(type) (!!((type) & MONKEY_UINPUT_TYPE_MASK))
+#define MONKEY_GET_DEV_TYPE(type)   (type & ~MONKEY_UINPUT_TYPE_MASK)
```
Comment (pkarashchenko) [inline testing/monkey/monkey_type.h:37]:
```suggestion
#define MONKEY_GET_DEV_TYPE(type)   ((type) & ~MONKEY_UINPUT_TYPE_MASK)
```

File: testing/monkey/monkey_type.h [hunk 22]
```diff
@@ -0,0 +1,118 @@
...
+{
+  struct monkey_config_s config;
+  enum monkey_mode_e mode;
+  FAR struct monkey_dev_s *devs[MONKEY_DEV_MAX_NUM];
+  int dev_num;
+  FAR struct monkey_recorder_s *recorder;
+  struct
+  {
+      struct monkey_dev_state_s state;
+      uint32_t time_stamp;
```
Comment (pkarashchenko) [inline testing/monkey/monkey_type.h:114]:
```suggestion
    struct monkey_dev_state_s state;
    uint32_t time_stamp;
```

File: testing/monkey/monkey_utils.h [hunk 23]
```diff
@@ -0,0 +1,94 @@
...
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
+ * WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.  See the
+ * License for the specific language governing permissions and limitations
+ * under the License.
+ *
+ ****************************************************************************/
+
+#ifndef __MONKEY_UTILS_H__
+#define __MONKEY_UTILS_H__
```
Comment (pkarashchenko) [inline testing/monkey/monkey_utils.h:22]: ditto

File: testing/monkey/monkey.c (hunk 2 above)
Comment (FASTSHIFT) [reply testing/monkey/monkey.c]: done

File: testing/monkey/monkey.c (hunk 1 above)
Comment (FASTSHIFT) [reply testing/monkey/monkey.c]: done

File: testing/monkey/monkey.c (hunk 3 above)
Comment (FASTSHIFT) [reply testing/monkey/monkey.c]: done

File: testing/monkey/monkey.c (hunk 4 above)
Comment (FASTSHIFT) [reply testing/monkey/monkey.c]: done

File: testing/monkey/monkey.h (hunk 5 above)
Comment (FASTSHIFT) [reply testing/monkey/monkey.h]: done

File: testing/monkey/monkey_assert.h (hunk 6 above)
Comment (FASTSHIFT) [reply testing/monkey/monkey_assert.h]: done

File: testing/monkey/monkey_dev.c (hunk 7 above)
Comment (FASTSHIFT) [reply testing/monkey/monkey_dev.c]: done

File: testing/monkey/monkey_dev.c (hunk 8 above)
Comment (FASTSHIFT) [reply testing/monkey/monkey_dev.c]: done

File: testing/monkey/monkey_dev.c (hunk 9 above)
Comment (FASTSHIFT) [reply testing/monkey/monkey_dev.c]: done

File: testing/monkey/monkey_dev.c (hunk 10 above)
Comment (FASTSHIFT) [reply testing/monkey/monkey_dev.c]: done

File: testing/monkey/monkey_dev.c (hunk 11 above)
Comment (FASTSHIFT) [reply testing/monkey/monkey_dev.c]: done

File: testing/monkey/monkey_dev.c (hunk 12 above)
Comment (FASTSHIFT) [reply testing/monkey/monkey_dev.c]: done

File: testing/monkey/monkey_dev.h (hunk 13 above)
Comment (FASTSHIFT) [reply testing/monkey/monkey_dev.h]: done

File: testing/monkey/monkey_log.c (hunk 14 above)
Comment (FASTSHIFT) [reply testing/monkey/monkey_log.c]: done

File: testing/monkey/monkey_main.c (hunk 16 above)
Comment (FASTSHIFT) [reply testing/monkey/monkey_main.c]: done

File: testing/monkey/monkey_recorder.c (hunk 17 above)
Comment (FASTSHIFT) [reply testing/monkey/monkey_recorder.c]: done

File: testing/monkey/monkey_recorder.c (hunk 18 above)
Comment (FASTSHIFT) [reply testing/monkey/monkey_recorder.c]: done

File: testing/monkey/monkey_recorder.h (hunk 19 above)
Comment (FASTSHIFT) [reply testing/monkey/monkey_recorder.h]: done

File: testing/monkey/monkey_type.h (hunk 21 above)
Comment (FASTSHIFT) [reply testing/monkey/monkey_type.h]: done

File: testing/monkey/monkey_type.h (hunk 20 above)
Comment (FASTSHIFT) [reply testing/monkey/monkey_type.h]: done

File: testing/monkey/monkey_type.h (hunk 22 above)
Comment (FASTSHIFT) [reply testing/monkey/monkey_type.h]: done

File: testing/monkey/monkey_utils.h (hunk 23 above)
Comment (FASTSHIFT) [reply testing/monkey/monkey_utils.h]: done

File: testing/monkey/monkey_dev.c [hunk 24]
```diff
@@ -0,0 +1,325 @@
...
+/****************************************************************************
+ * Name: button_read
+ ****************************************************************************/
+
+static bool button_read(int fd, FAR uint32_t *value)
+{
+  btn_buttonset_t buttonset;
+
+  int ret = read(fd, &buttonset, sizeof(buttonset));
+  if (ret < 0)
```
Comment (pkarashchenko) [inline testing/monkey/monkey_dev.c:128]:
```suggestion
  if (ret != sizeof(buttonset))
```

File: testing/monkey/monkey_dev.c [hunk 25]
```diff
@@ -0,0 +1,325 @@
...
+
+/****************************************************************************
+ * Name: monkey_dev_get_available
+ ****************************************************************************/
+
+int monkey_dev_get_available(FAR struct monkey_dev_s *devs[], int dev_num)
+{
+  int i;
+  int available = 0;
+  struct pollfd fds[MONKEY_DEV_MAX_NUM];
```
Comment (pkarashchenko) [inline testing/monkey/monkey_dev.c:299]: Do we need `memset(&fds, 0, sizeof(fds));`?

**Comment by pkarashchenko** [comment]
In general looks good

File: testing/monkey/monkey_dev.c (hunk 24 above)
Comment (FASTSHIFT) [reply testing/monkey/monkey_dev.c]: done

File: testing/monkey/monkey_dev.c (hunk 25 above)
Comment (FASTSHIFT) [reply testing/monkey/monkey_dev.c:299]: done

**Review by pkarashchenko — APPROVED** [review]

**Pull request merged**